### PR TITLE
BTreeMap: reuse BoxedNode instances directly instead of their contents 

### PR DESCRIPTION
**Pull request by ssomers**
r? @Mark-Simulacrum

### CODE REVIEW

File: library/alloc/src/collections/btree/node.rs [hunk 1]
```diff
@@ -215,10 +225,10 @@ impl<K, V> Root<K, V> {
 
         let top = self.node.ptr;
 
+        let mut internal_node = unsafe { self.internal_node_as_mut() };
+        let internal_node = internal_node.as_internal_mut();
         self.node = unsafe {
-            BoxedNode::from_ptr(
-                self.node_as_mut().cast_unchecked::<marker::Internal>().first_edge().descend().node,
-            )
+            mem::replace(&mut internal_node.edges[0], MaybeUninit::uninit()).assume_init()
```
Comment (Mark-Simulacrum) [inline library/alloc/src/collections/btree/node.rs:231]: This feels like we're (potentially) doing needless work -- this looks equivalent to `internal_node.edges[0].assume_init_read()` which avoids the spurious write to edges -- can we use that instead?

(I suspect in practice LLVM probably ignores writing uninitialized data to some region of memory, but not sure).

File: library/alloc/src/collections/btree/node.rs [hunk 2]
```diff
@@ -185,20 +197,18 @@ impl<K, V> Root<K, V> {
     /// make that new node the root node, and return it. This increases the height by 1
     /// and is the opposite of `pop_internal_level`.
     pub fn push_internal_level(&mut self) -> NodeRef<marker::Mut<'_>, K, V, marker::Internal> {
+        let top = mem::replace(&mut self.node, BoxedNode { ptr: Unique::dangling() });
         let mut new_node = Box::new(unsafe { InternalNode::new() });
-        new_node.edges[0].write(unsafe { BoxedNode::from_ptr(self.node.as_ptr()) });
+        new_node.edges[0].write(top);
 
         self.node = BoxedNode::from_internal(new_node);
```
[the remark this answers is not in the document]
Comment (ssomers) [reply library/alloc/src/collections/btree/node.rs:204]: I don't think it's just noise. I always get consistent results within 2%, provided I run old and new on the same system, so without rebooting. Sometimes some kinds of test win and others lose, but not here.

That said, the new version coming up is more convincing:
```
benchcmp old new --threshold 3
 name                                         old ns/iter  new ns/iter  diff ns/iter   diff %  speedup
 btree::map::clone_slim_100_and_clear         2,574        2,083                -491  -19.08%   x 1.24
 btree::map::clone_slim_100_and_drain_all     3,927        3,472                -455  -11.59%   x 1.13
 btree::map::clone_slim_100_and_drain_half    3,337        2,832                -505  -15.13%   x 1.18
 btree::map::clone_slim_100_and_into_iter     2,597        2,075                -522  -20.10%   x 1.25
 btree::map::clone_slim_100_and_pop_all       3,751        3,260                -491  -13.09%   x 1.15
 btree::map::clone_slim_100_and_remove_all    4,548        4,079                -469  -10.31%   x 1.11
 btree::map::clone_slim_100_and_remove_half   3,404        2,840                -564  -16.57%   x 1.20
```

File: library/alloc/src/collections/btree/node.rs (hunk 1 above)
Comment (ssomers) [reply library/alloc/src/collections/btree/node.rs:231]: Indeed and I did the logical equivalent ptr::read on the other side.

**Comment by ssomers** [comment]
Swapped the 2 commits here and based on #77471 to avoid a benign conflict

**Comment by ssomers** [comment]
Nothing wrong, just easier to follow if I do the first commit separately.

**Comment by ssomers** [comment]
Still showing the same improvement, but now with the opposite on sets.

```benchcmp old new --threshold 3
 name                                         old ns/iter  new ns/iter  diff ns/iter   diff %  speedup
 btree::map::clone_slim_100                   2,043        2,106                  63    3.08%   x 0.97
 btree::map::clone_slim_100_and_clear         2,551        2,065                -486  -19.05%   x 1.24
 btree::map::clone_slim_100_and_drain_all     3,891        3,441                -450  -11.57%   x 1.13
 btree::map::clone_slim_100_and_drain_half    3,351        2,844                -507  -15.13%   x 1.18
 btree::map::clone_slim_100_and_into_iter     2,587        2,045                -542  -20.95%   x 1.27
 btree::map::clone_slim_100_and_pop_all       3,684        3,189                -495  -13.44%   x 1.16
 btree::map::clone_slim_100_and_remove_all    4,515        4,044                -471  -10.43%   x 1.12
 btree::map::clone_slim_100_and_remove_half   3,388        2,823                -565  -16.68%   x 1.20
 btree::set::clone_100_and_remove_all         3,397        3,792                 395   11.63%   x 0.90
 btree::set::clone_10k_and_remove_all         400,900      436,475            35,575    8.87%   x 0.92
```

**Comment by Mark-Simulacrum** [comment]
Oh, hm, I just saw that #78104 lists itself as a better alternative to this. @bors r-

**Comment by ssomers** [comment]
Well I actually rebased #77244 anticepating this, so you approved it again…

It's mostly a logical alternative, explicitly not reusing rather than explicitly reusing, but it's easy enough to let #78104 overwrite this little change.

**Comment by Mark-Simulacrum** [comment]
Okay. I don't mind the extra PR. @bors r+

**Comment by ssomers** [comment]
#77244 was rebased on the same changes and made it through first. I wondered what bors would do and as far as I can tell it just turned away in disgust. The commit in #77244 is a different SHA-1 though, because it was rebased too.

**Pull request closed without merging**